### PR TITLE
Contact mechanics (conforming) input files

### PR DESCRIPTION
**Pull request by castelletto1**
This PR corrects some input files involving the contact solver for conforming mesh (Lagrange multipliers) that were producing singular Jacobian matrices. In all cases, the singularity appears at the first iteration under the assumption of stick conditions. The original files have all nodes in the `z` direction constrained -- the internally generated mesh has `nz` equal to 1. As a consequence, the tangential traction component in the `z` direction can only be computed up to constant, the Jacobian matrix having a null eigenvalue. The proposed fix consists in adding a layer of cell in `z`, i.e `nz = 2`. The problem has been exposed during the testing phase for PR #2881. Note that even using develop the issue can be reproduced. For example, in the [Sneddon](https://github.com/GEOS-DEV/GEOS/blob/develop/inputFiles/lagrangianContactMechanics/Sneddon_contactMechanics_smoke.xml) smoke test, it is enough to change the [mesh](https://github.com/GEOS-DEV/GEOS/blob/3cdbde87c3ca7d356e4d70ad47faa3e383ed872f/inputFiles/lagrangianContactMechanics/Sneddon_contactMechanics_smoke.xml#L11-L20) as follows:

```xml
  <Mesh>
    <InternalMesh
      name="mesh1"
      elementTypes="{ C3D8 }"
      xCoords="{ -20, -2, 2, 20 }"
      yCoords="{ -20, -2, 2, 20 }"
      zCoords="{ 0, 1 }"
      nx="{ 4, 22, 4 }"
      ny="{ 4, 21, 4 }"  // <-- 21 instead of 22
      nz="{ 1 }"
      cellBlockNames="{ cb1 }"/>
  </Mesh>
```

@jhuang2601 could you please confirm that there are no implications on the tutorials (for example Sneddon)? Thank you.

Also, the `ExponentialDecayPermeability_conformingFracture` smoke test had no displacement constraint in the `y` direction. I added it taking advantage of symmetry for two lines of nodes. The only requirement is that the mesh must have an even number of cell in the `y` direction to allow for selecting all nodes at `(xmin, 0)` and `(xmax, 0)`.

Requires new baselines: GEOS-DEV/integratedTests/pull/90

### CODE REVIEW

**Review by CusiniM — APPROVED** [review]

**Comment by jhuang2601** [comment]
@castelletto1 I can reproduce the issue for running the `Sneddon_contactMechanics_smoke.xml` with the suggested change of mesh:
```
Time: 0.00e+00 s, dt: 1 s, Cycle: 0

  Number of element for each fracture state: stick:            9 | slip:             0 | open:             0
    Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
        ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
        lagrangiancontact: Global solution scaling factor = 1
    Attempt:  0, ConfigurationIter:  0, NewtonIter:  1
        ( Rdisplacement, Rtraction, Rtotal ) = ( -nan           , -nan           , -nan            )        ( R ) = ( -nan )
        Last LinSolve(iter,res) = (   1, -nan )
```

**Comment by CusiniM** [comment]
> @castelletto1 I can reproduce the issue for running the `Sneddon_contactMechanics_smoke.xml` with the suggested change of mesh:
> 
> ```
> Time: 0.00e+00 s, dt: 1 s, Cycle: 0
> 
>   Number of element for each fracture state: stick:            9 | slip:             0 | open:             0
>     Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
>         ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
>         lagrangiancontact: Global solution scaling factor = 1
>     Attempt:  0, ConfigurationIter:  0, NewtonIter:  1
>         ( Rdisplacement, Rtraction, Rtotal ) = ( -nan           , -nan           , -nan            )        ( R ) = ( -nan )
>         Last LinSolve(iter,res) = (   1, -nan )
> ```

The change should have fixed an issue not introduced one.

**Comment by CusiniM** [comment]
> > @castelletto1 I can reproduce the issue for running the `Sneddon_contactMechanics_smoke.xml` with the suggested change of mesh:
> > ```
> > Time: 0.00e+00 s, dt: 1 s, Cycle: 0
> > 
> >   Number of element for each fracture state: stick:            9 | slip:             0 | open:             0
> >     Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
> >         ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
> >         lagrangiancontact: Global solution scaling factor = 1
> >     Attempt:  0, ConfigurationIter:  0, NewtonIter:  1
> >         ( Rdisplacement, Rtraction, Rtotal ) = ( -nan           , -nan           , -nan            )        ( R ) = ( -nan )
> >         Last LinSolve(iter,res) = (   1, -nan )
> > ```
> 
> The change should have fixed an issue not introduced one.

OH sorry, I see what you mean. I think the question is whether the match to the analytical solution is significantly impacted by adding an extra element in the z-direction.

**Comment by jhuang2601** [comment]
@castelletto1 However, with updated `Sneddon_contactMechanics_benchmark.xml` in mesh settings `nz="{ 2 }`, the case cannot be running with develop repo anymore and always report memory issue. Tested with 1, 4, 8 ranks and issue persists.
```
Time: 0.00e+00 s, dt: 1 s, Cycle: 0

  Number of element for each fracture state: stick:          220 | slip:             0 | open:             0
    Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
        ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )

UMFPACK V5.7.9 (Oct 20, 2019): ERROR: out of memory

***** ERROR
***** LOCATION: /shared/data1/Users/j0551570/Compilation/Develop_010224/GEOS/src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp:158
***** Controlling expression (should be false): true
***** Rank 0: SuiteSparse: umfpack_dl_numeric failed.
```
If keeping the initial setting with  `nz="{ 1 }` for `Sneddon_contactMechanics_benchmark.xml`, the case is running and can match with analytical solution.

**Comment by jhuang2601** [comment]
> > > @castelletto1 I can reproduce the issue for running the `Sneddon_contactMechanics_smoke.xml` with the suggested change of mesh:
> > > ```
> > > Time: 0.00e+00 s, dt: 1 s, Cycle: 0
> > > 
> > >   Number of element for each fracture state: stick:            9 | slip:             0 | open:             0
> > >     Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
> > >         ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
> > >         lagrangiancontact: Global solution scaling factor = 1
> > >     Attempt:  0, ConfigurationIter:  0, NewtonIter:  1
> > >         ( Rdisplacement, Rtraction, Rtotal ) = ( -nan           , -nan           , -nan            )        ( R ) = ( -nan )
> > >         Last LinSolve(iter,res) = (   1, -nan )
> > > ```
> > 
> > 
> > The change should have fixed an issue not introduced one.
> 
> OH sorry, I see what you mean. I think the question is whether the match to the analytical solution is significantly impacted by adding an extra element in the z-direction.

@CusiniM Try to confirm two things: 1) adding an extra element can fix the issue for smoke test; 2) same solution seems not applicable for the benchmark example.

**Comment by CusiniM** [comment]
> @castelletto1 However, with updated `Sneddon_contactMechanics_benchmark.xml` in mesh settings `nz="{ 2 }`, the case cannot be running with develop repo anymore and always report memory issue. Tested with 1, 4, 8 ranks and issue persists.
> 
> ```
> Time: 0.00e+00 s, dt: 1 s, Cycle: 0
> 
>   Number of element for each fracture state: stick:          220 | slip:             0 | open:             0
>     Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
>         ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
> 
> UMFPACK V5.7.9 (Oct 20, 2019): ERROR: out of memory
> 
> ***** ERROR
> ***** LOCATION: /shared/data1/Users/j0551570/Compilation/Develop_010224/GEOS/src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp:158
> ***** Controlling expression (should be false): true
> ***** Rank 0: SuiteSparse: umfpack_dl_numeric failed.
> ```
> 
> If keeping the initial setting with `nz="{ 1 }` for `Sneddon_contactMechanics_benchmark.xml`, the case is running and can match with analytical solution.


I see that's a linear solver issue. I guess the benchmark is now too large for the direct solver. @jhuang2601  can you change the `solverType` to `gmres` with `preconditonerType='mgr'`?  
Although we should check that mgr strategy.

**Comment by jhuang2601** [comment]
> > @castelletto1 However, with updated `Sneddon_contactMechanics_benchmark.xml` in mesh settings `nz="{ 2 }`, the case cannot be running with develop repo anymore and always report memory issue. Tested with 1, 4, 8 ranks and issue persists.
> > ```
> > Time: 0.00e+00 s, dt: 1 s, Cycle: 0
> > 
> >   Number of element for each fracture state: stick:          220 | slip:             0 | open:             0
> >     Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
> >         ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
> > 
> > UMFPACK V5.7.9 (Oct 20, 2019): ERROR: out of memory
> > 
> > ***** ERROR
> > ***** LOCATION: /shared/data1/Users/j0551570/Compilation/Develop_010224/GEOS/src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp:158
> > ***** Controlling expression (should be false): true
> > ***** Rank 0: SuiteSparse: umfpack_dl_numeric failed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If keeping the initial setting with `nz="{ 1 }` for `Sneddon_contactMechanics_benchmark.xml`, the case is running and can match with analytical solution.
> 
> I see that's a linear solver issue. I guess the benchmark is now too large for the direct solver. @jhuang2601 can you change the `solverType` to `gmres` with `preconditonerType='mgr'`? Although we should check that mgr strategy.

@CusiniM I can give it a try with iterative solver. But the existing case does not complain the current mesh settings, do we still have to update the mesh?

**Comment by CusiniM** [comment]
> > > @castelletto1 However, with updated `Sneddon_contactMechanics_benchmark.xml` in mesh settings `nz="{ 2 }`, the case cannot be running with develop repo anymore and always report memory issue. Tested with 1, 4, 8 ranks and issue persists.
> > > ```
> > > Time: 0.00e+00 s, dt: 1 s, Cycle: 0
> > > 
> > >   Number of element for each fracture state: stick:          220 | slip:             0 | open:             0
> > >     Attempt:  0, ConfigurationIter:  0, NewtonIter:  0
> > >         ( Rdisplacement, Rtraction, Rtotal ) = (    1.000000e+00,    1.000000e+00,    1.000000e+00 )        ( R ) = ( 1.00e+00 )
> > > 
> > > UMFPACK V5.7.9 (Oct 20, 2019): ERROR: out of memory
> > > 
> > > ***** ERROR
> > > ***** LOCATION: /shared/data1/Users/j0551570/Compilation/Develop_010224/GEOS/src/coreComponents/linearAlgebra/interfaces/direct/SuiteSparse.cpp:158
> > > ***** Controlling expression (should be false): true
> > > ***** Rank 0: SuiteSparse: umfpack_dl_numeric failed.
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > If keeping the initial setting with `nz="{ 1 }` for `Sneddon_contactMechanics_benchmark.xml`, the case is running and can match with analytical solution.
> > 
> > 
> > I see that's a linear solver issue. I guess the benchmark is now too large for the direct solver. @jhuang2601 can you change the `solverType` to `gmres` with `preconditonerType='mgr'`? Although we should check that mgr strategy.
> 
> @CusiniM I can give it a try with iterative solver. But the existing case does not complain the current mesh settings, do we still have to update the mesh?

Yes, I think that it's luck that it does not fail. The problem is ill-posed and the system is singular. I was looking at the mgr strategy though and I think it needs to be fixed. I will push a fix in this branch.

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/LagrangianContactMechanics.hpp [hunk 1]
```diff
@@ -58,20 +57,23 @@ class LagrangianContactMechanics : public MGRStrategyBase< 1 >
...
     setupLabels();
 
     // Level 0
     m_levelFRelaxType[0]          = MGRFRelaxationType::none;
-    m_levelInterpType[0]          = MGRInterpolationType::jacobi;
-    m_levelRestrictType[0]        = MGRRestrictionType::injection;
-    m_levelCoarseGridMethod[0]    = MGRCoarseGridMethod::galerkin;
+    m_levelFRelaxIters[0]         = 0;
+
     m_levelGlobalSmootherType[0]  = MGRGlobalSmootherType::blockJacobi;
```
Comment (CusiniM) [inline src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/LagrangianContactMechanics.hpp]: @castelletto1 I have tried removing the global smoother and using `jacobi` as the f-relaxation but the LS does not converge once the fracture opens. It seems to work fine for the closed case.

Comment (jhuang2601) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/LagrangianContactMechanics.hpp]: > @castelletto1 I have tried removing the global smoother and using `jacobi` as the f-relaxation but the LS does not converge once the fracture opens. It seems to work fine for the closed case.

@CusiniM With your fix on MGR, I can run two tutorial examples (Sneddon and TFrac) with updated mesh. Both cases can match with the analytical solution and I did not see any convergence issue with open fracture.

**Review by jhuang2601 — APPROVED** [review]

File: src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/LagrangianContactMechanics.hpp (hunk 1 above)
Comment (castelletto1) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/LagrangianContactMechanics.hpp]: I'll have a quick look at the MGR strategy. If it works, I suggest we keep it as is, and maybe improve it in future PRs.

Comment (castelletto1) [reply src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/LagrangianContactMechanics.hpp]: As a general comment, we should rely on the serial direct solver for smoke tests exclusively. Benchmark tests should always consider an iterative solver (or the parallel direct solver).

**Comment by castelletto1** [comment]
Is there a reason why

- `ContactMechanics_slippingFault_horizontal_smoke.xml`
- `ContactMechanics_slippingFault_vertical_smoke.xml`

are not not included in the integrated tests? Since we need to do a rebaseline, if you think there is value, I can add them.

**Comment by CusiniM** [comment]
> Is there a reason why
> 
>     * `ContactMechanics_slippingFault_horizontal_smoke.xml`
> 
>     * `ContactMechanics_slippingFault_vertical_smoke.xml`
> 
> 
> are not not included in the integrated tests? Since we need to do a rebaseline, if you think there is value, I can add them.

I don't really see a good reason for that. I think we should add them.

**Pull request merged**